pkg/mdns/setup.go

```diff
@@ -23,14 +25,30 @@ func setup(c *caddy.Controller) error {
 		return plugin.Error("mdns", c.ArgErr())
 	}
 
+	// Because the plugin interface uses a value receiver, we need to make these
+	// pointers so all copies of the plugin point at the same maps.
+	mdnsHosts := make(map[string]*mdns.ServiceEntry)
+	srvHosts := make(map[string][]*mdns.ServiceEntry)
+	m := MDNS{Domain: domain, mdnsHosts: &mdnsHosts, srvHosts: &srvHosts}
+
 	c.OnStartup(func() error {
-		//once.Do(func() { metrics.MustRegister(c, requestCount) })
+		go browseLoop(&m)
 		return nil
 	})
 
 	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
-		return MDNS{Next: next, Domain: domain}
+		m.Next = next
+		return m
 	})
 
 	return nil
 }
+
+func browseLoop(m *MDNS) {
+	for {
+		m.BrowseMDNS()
+		// The OpenShift Corefile configures caching for 30 seconds, so there's little
+		// point in updating more often than that.
+		time.Sleep(30 * time.Second)
```

Comment:
We should check if it is possible to retrieve that value. It sounds like it should not. If it is indeed not possible, it would be good to have this as a config option rather than having it hardcoded.

Reply:
Does this mean that when a new worker starts announcing its _workstation, that it may take 30+ seconds before we are able to resolve it?

Reply:
It's possible. If an A request is made and the upstream DNS server responds with something, that response may be cached. For example, when I try to query a not-yet-running ostest.metalkube.com address I get back an SOA record from upstream somewhere and that gets cached, which masks the actual address until the cache expires.

It's also an issue for SRV records. When I query for SRV, whatever nodes are running at that time get cached and any new ones don't show up until the cache expires. However, in that case I can mitigate it by shortening our SRV TTL, which limits the window where there is inconsistency. I can't do the same for A records because we aren't actually serving the initial response in that case.

I think the only way to get rid of it completely would be to disable caching by moving the plugin back above cache in the plugin list. Given that we're now running mdns in the background that might be okay since our responses are all just map lookups.

I'll add a commit to shorten the TTLs since we probably want to do that anyway, and then we can decide if we still need to kill the cache too.